docs/javascript-unit-tests.md

```diff
@@ -0,0 +1,712 @@
+which can be used to performing actions
+that are needed before and after running all tests or each test in a suite.
+For example, the tests in
+[`Analytics-spec.js`](https://github.com/cfpb/cfgov-refresh/blob/add-js-unit-testing-doc/test/unit_tests/js/modules/Analytics-spec.js#L7-L32)
+use both `beforeAll` and `beforeEach` inside the root `describe` block to
+do a variable definition for all tests at the beginning of the suite
+and reset the `dataLayer` before each test, respectively.
+
+[Check out the Jest documentation on “Setup and teardown” methods.](https://jestjs.io/docs/en/setup-teardown)
+
+A common structure when the DOM is involved
+is to create a constant representing an HTML snippet to test,
+then – in a `beforeEach` or `beforeAll`
+(depending on whether the tests modify the markup or not) –
+set `document.body.innerHTML` to that snippet.
+
+Use `beforeAll` to attach HTML markup that is unaffected by the tests,
+e.g., the tests in
+[`footer-button-spec.js`](https://github.com/cfpb/cfgov-refresh/blob/add-js-unit-testing-doc/test/unit_tests/js/modules/footer-button-spec.js).
+Use `beforeEach` to reset manipulated markup between tests, e.g., the tests in
+[`Notification-spec.js`](https://github.com/cfpb/cfgov-refresh/blob/add-js-unit-testing-doc/test/unit_tests/js/molecules/Notification-spec.js).
+
+[See “Testing DOM manipulation”](#testing-dom-manipulation)
+in the “Common test patterns” section of this page
+for a more in-depth discussion of this scenario.
+
+
+
+## Common test patterns
+
+
+### Testing a basic function
+
+Testing simple functions is pretty straightforward.
+Each function in a module should have tests set up as
+a child `describe` within the module’s own `describe`.
+Then, write a number of `it` statements in prose that describe
+how the function should respond to various kinds of input.
+Inside each `it`, invoke the function
+with the input described in the `it` statement
+and use `expect` to check that you receive the desired result.
+
+Here is a simple example from our
+[array helpers module](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/unprocessed/js/modules/util/array-helpers.js)
+(`cfgov/unprocessed/js/modules/util/array-helpers.js`):
+
+```js
+function indexOfObject( array, key, val ) {
+  let match = -1;
+
+  if ( !array.length > 0 ) {
+    return match;
+  }
+
+  array.forEach( function( item, index ) {
+    if ( item[key] === val ) {
+      match = index;
+    }
+  } );
+
+  return match;
+}
+```
+
+Tests for that function, from
+[test/unit_tests/js/modules/util/array-helpers-spec.js](https://github.com/cfpb/cfgov-refresh/blob/master/test/unit_tests/js/modules/util/array-helpers-spec.js):
+
+```js
+describe( 'indexOfObject()', () => {
+  it( 'should return -1 if the array is empty', () => {
+    array = [];
```

Comment:
Should be declared with `const array = [];`. Is this not declared in the source?

Reply:
Here and below.

Reply:
Similar to the previous comment on the `sample-spec.js` excerpt, it's declared at the top of the source file, with `let`, as it changes for each test in the file.

Reply:
I defer to your judgement on whether the excerpt should declare it or not.